gitlint/__init__.py

```diff
@@ -217,49 +237,49 @@ def main(argv, stdout=sys.stdout, stderr=sys.stderr):
     gitlint_config = get_config(repository_root)
     json_result = {}
 
-    for filename in sorted(modified_files.keys()):
-        rel_filename = os.path.relpath(filename)
-        if not json_output:
-            stdout.write('Linting file: %s%s' %
-                         (termcolor.colored(rel_filename, attrs=('bold',)),
-                          linesep))
-        if arguments['--force']:
-            modified_lines = None
-        else:
-            modified_lines = vcs.modified_lines(filename,
-                                                modified_files[filename],
-                                                commit=commit)
-
-        result = linters.lint(
-            filename, modified_lines, gitlint_config)
-        result = result[filename]
-
-        output_lines = []
-        if result.get('error'):
-            output_lines.extend(
-                '%s: %s' % (ERROR, reason) for reason in result.get('error')
-            )
-            linter_not_found = True
-        if result.get('skipped'):
-            output_lines.extend(
-                '%s: %s' % (SKIPPED, reason) for reason in result.get('skipped')
-            )
-        if result.get('comments', []) == []:
-            if not output_lines:
-                output_lines.append(OK)
-        else:
-            files_with_problems += 1
-            for data in result['comments']:
-                formatted_message = format_comment(data)
-                output_lines.append(formatted_message)
-                data['formatted_message'] = formatted_message
-
-        if json_output:
-            json_result[filename] = result
-        else:
-            output = linesep.join(output_lines)
-            stdout.write(output)
-            stdout.write(linesep + linesep)
+    with futures.ThreadPoolExecutor(max_workers=cpu_count()) as executor:
+        future_to_file = {executor.submit(process_file, vcs, filename,
+                                          modified_files[filename], commit,
+                                          arguments['--force'], gitlint_config):
+                          filename for filename in sorted(modified_files.keys())
+                          }
+
+        for future in futures.as_completed(future_to_file):
+            result = future.result()
+            filename = future_to_file[future]
+            rel_filename = os.path.relpath(filename)
+
+            if not json_output:
+                stdout.write('Linting file: %s%s' %
```

Comment:
This is tricky. Now the message is no longer "Linting file ..." as the action already finished. Additionally you loose the ability to track whether the system is stuck in a particular file or not and that the files are returned in no consistent order (i.e. different runs will have different output).

An alternative is to use [executor.map](http://pythonhosted.org/futures/#concurrent.futures.Executor.map). That allows to keep the current semantics and actually simplifies the code as you can get rid of the `future_to_file` mapping.